### PR TITLE
RDM-9556: Data Store Changes for Conditional Event Post State

### PR DESCRIPTION
**Pull request by kiran-yenigala-hmcts**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RDM-9556

### Change description ###

Post State Condition changes

**Does this PR introduce a breaking change?** (check one with "x")

```
[X ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/ccd/domain/service/common/CaseStateUpdateService.java [hunk 1]
```diff
@@ -3,34 +3,43 @@
...
 import javax.inject.Inject;
 import org.springframework.stereotype.Service;
-import uk.gov.hmcts.ccd.domain.casestate.EnablingConditionSorter;
+import uk.gov.hmcts.ccd.domain.enablingcondition.PrioritiseEnablingCondition;
 import uk.gov.hmcts.ccd.domain.model.definition.CaseDetails;
 import uk.gov.hmcts.ccd.domain.model.definition.CaseEventDefinition;
 import uk.gov.hmcts.ccd.domain.model.definition.EventPostStateDefinition;
 
 @Service
 public class CaseStateUpdateService {
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/common/CaseStateUpdateService.java:16]:
```suggestion
public class CasePostStateService {
```
this class is not really doing any case update itself right?

File: src/main/java/uk/gov/hmcts/ccd/domain/service/common/CaseStateUpdateService.java [hunk 2]
```diff
@@ -3,34 +3,43 @@
...
+        this.prioritiseEnablingCondition = prioritiseEnablingCondition;
+        this.casePostStateEvaluationService = casePostStateEvaluationService;
     }
 
-    public Optional<String> retrieveCaseState(CaseEventDefinition caseEventDefinition, CaseDetails caseDetails) {
-        List<EventPostStateDefinition> eventPostStateDefinitions = caseEventDefinition.getPostStates();
-        this.enablingConditionSorter.sortEventPostStates(eventPostStateDefinitions);
-        Map<String, JsonNode> caseEventData = caseDetails.getCaseEventData(caseEventDefinition);
-        Optional<String> postStateReference = this.stateReferenceService
+    public String retrieveCaseState(CaseEventDefinition caseEventDefinition, CaseDetails caseDetails) {
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/common/CaseStateUpdateService.java:28]:
```suggestion
    public String evaluateCaseState(CaseEventDefinition caseEventDefinition, CaseDetails caseDetails) {
```

Comment (mario-paniccia) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/common/CaseStateUpdateService.java:28]: or `evaluateCaseState`?

File: src/main/java/uk/gov/hmcts/ccd/domain/service/createevent/CreateCaseEventService.java [hunk 3]
```diff
@@ -234,9 +234,9 @@ private void mergeUpdatedFieldsToCaseDetails(final Map<String, JsonNode> data,
     }
 
     private void updateCaseState(CaseDetails caseDetails, CaseEventDefinition caseEventDefinition) {
-        Optional<String> postState = caseStateUpdateService.retrieveCaseState(caseEventDefinition, caseDetails);
-        if (postState.isPresent() && !equalsIgnoreCase(CaseStateDefinition.ANY, postState.get())) {
-            caseDetails.setState(postState.get());
+        String postState = caseStateUpdateService.retrieveCaseState(caseEventDefinition, caseDetails);
+        if (!equalsIgnoreCase(CaseStateDefinition.ANY, postState)) {
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/createevent/CreateCaseEventService.java:238]: if(!shouldRemainOnCurrentState(postState)){

where `shouldRemainOnCurrentState(String state)` is a private method containing `equalsIgnoreCase(CaseStateDefinition.ANY, postState)`

File: src/main/java/uk/gov/hmcts/ccd/domain/enablingcondition/jexl/JexlEnablingConditionParser.java [hunk 4]
```diff
@@ -1,4 +1,4 @@
-package uk.gov.hmcts.ccd.domain.casestate.jexl;
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/enablingcondition/jexl/JexlEnablingConditionParser.java:1]: are we able to support enabling condition on null fields? And on blank String fields? what is the notation that we need to use? same for both?
for example something like this what would match:
`claimantIndType.claimant_title1 ==""`

File: src/main/java/uk/gov/hmcts/ccd/domain/service/createevent/CreateCaseEventService.java (hunk 3 above)
Comment (mario-paniccia) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/createevent/CreateCaseEventService.java:238]: suggesting also the same refactor on CreateCaseEventService

File: src/main/java/uk/gov/hmcts/ccd/domain/enablingcondition/EnablingConditionFormatter.java [hunk 5]
```diff
@@ -1,4 +1,4 @@
-package uk.gov.hmcts.ccd.domain.casestate;
+package uk.gov.hmcts.ccd.domain.enablingcondition;
 
 public interface EnablingConditionFormatter {
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/enablingcondition/EnablingConditionFormatter.java:3]:
```suggestion
public interface EnablingConditionConverter {
```
also can you add a small Javadoc explaining this is converting our internal enabling condition syntax to the Jexl one?

Comment (mario-paniccia) [reply src/main/java/uk/gov/hmcts/ccd/domain/enablingcondition/EnablingConditionFormatter.java:3]: and change the method:
    `String format(String enablingCondition);`
`String convertToJexl(String enablingCondition);`

Comment (mario-paniccia) [reply src/main/java/uk/gov/hmcts/ccd/domain/enablingcondition/EnablingConditionFormatter.java:3]: sorry the javadoc should be on the JexlConverter subclass not on the interface. On the interface we can add a more generic description of the class withouth mentioning Jexl

File: src/main/java/uk/gov/hmcts/ccd/domain/enablingcondition/jexl/JexlEnablingConditionFormatter.java [hunk 6]
```diff
@@ -1,4 +1,4 @@
-package uk.gov.hmcts.ccd.domain.casestate.jexl;
+package uk.gov.hmcts.ccd.domain.enablingcondition.jexl;
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/enablingcondition/jexl/JexlEnablingConditionFormatter.java:1]: shall we rename to JexlEnablingConditionConverter ?
and add a little Javadoc

File: src/main/java/uk/gov/hmcts/ccd/domain/enablingcondition/jexl/JexlEnablingConditionParser.java (hunk 4 above)
Comment (mario-paniccia) [reply src/main/java/uk/gov/hmcts/ccd/domain/enablingcondition/jexl/JexlEnablingConditionParser.java:1]: actually can we have a unite test for the scenarios above?

File: src/main/java/uk/gov/hmcts/ccd/domain/enablingcondition/EnablingConditionFormatter.java (hunk 5 above)
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/ccd/domain/enablingcondition/EnablingConditionFormatter.java:3]: done

File: src/main/java/uk/gov/hmcts/ccd/domain/enablingcondition/jexl/JexlEnablingConditionParser.java (hunk 4 above)
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/ccd/domain/enablingcondition/jexl/JexlEnablingConditionParser.java:1]: Null we can't check, as we are using Map, which not allowed to save nulls

**Review by mario-paniccia — APPROVED** [review]
👍

**Review by Olawale — APPROVED** [review]

**Pull request merged**